Update _P043_ClkOutput.ino

StefanNetzer opened the pull request:
Add additional events per schedule line

The clock - output plugin is an easy way to create an event scheduler. But, it does not provide events for each event, but only a generic on/off event.

With this change I have two values the schedule line # (variable schedule) and the event value (i.e. Temp).

I have changed the selector to an FormNumericBox, since this is more convienient to enter a value (i.e. Temperature)

Sensor
GPIO → Clock Event: - None -
Day,Time 1: All,00:00
Temp1: 0
Day,Time 2: All,00:00
Temp2: 0

With this one gets two output Values

Output
Temp

and of cause two events, which I may evaluate in the rules

556164: EVENT: timer#Output=1.00
556261: EVENT: timer#Temp=10.00

With this enhancement it is easy to use ESP as an edge device for tons of use cases ( thermostat, irrigation systems )

StefanNetzer committed:
Update _P043_ClkOutput.ino

StefanNetzer committed:
Merge branch 'mega' into mega

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -1,10 +1,9 @@
 //#######################################################################################################
 
-
+#include "_Plugin_Helper.h"
```
Make sure to have this specific include at the top line, before `#ifdef USES_P043` as those defines are set by including `_Plugin_Helper.h`

StefanNetzer replied:
Done

StefanNetzer replied:
Output Selector implemented, this should solve the issue.

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -28,7 +27,7 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
         Device[deviceCount].InverseLogicOption = false;
         Device[deviceCount].FormulaOption = false;
-        Device[deviceCount].ValueCount = 1;
+        Device[deviceCount].ValueCount = 2;
```
Changing value count may affect controllers like Domoticz.
So I have to have a closer look at this to see if this is an issue.

StefanNetzer replied:
Well, I am using FHEM, so I cannot comment on this one.

StefanNetzer replied:
changed the output_data_type to simple

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -68,10 +64,7 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
-          addHtml(" ");
-          byte choice = ExtraTaskSettings.TaskDevicePluginConfig[x];
-          addSelector(String(F("p043_state")) + (x), 3, options, NULL, NULL, choice);
+        	  addFormNumericBox(String(F("Temp")) + (x + 1), String(F("p043_state")) + (x), ExtraTaskSettings.TaskDevicePluginConfig[x],-100,100 );
```
Does this break existing configuration of units out there? 
What is the behavior if there is the "default" value? (assuming there is a default value and not random data)

Also what is stored in this "Temp" value?
Something temporary, or a temperature?

StefanNetzer replied:
Please give me a little bit of time. I will do a migration from the old to the suggested version, to identify issues here. The values old were default 0 (not active) 1 ( gpio off ) 2 (gpio on).  Temp"erature" is the label  for the value in my case desirerd temperature for a thermostat. 

I am not certain, if it is the best solution to use the value for both the old and the new use case, I am open to suggestions.

TD-er replied:
That was what I was afraid of, naming it with a very specific use case in mind.
Why not name it "Output2"?
Or if you need to have a specific label, then the label should also be configurable or maybe use the task variable name for it (as is already done in events when the call to `PLUGIN_READ` returns `success = true`.)

StefanNetzer replied:
renamed to Output2

StefanNetzer replied:
Naming changed to Output 2, thanks for this advice.  Tested the Upgrade from current version of  the plugin, to the proposed version. All values survived the update. Tested that the original schedule is still executed as before. Tested OK.

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -114,10 +107,13 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
-              UserVar[event->BaseVarIndex] = state;
+              byte state_relais = state -1;
+              if (state == 2 ) digitalWrite(CONFIG_PIN1, state_relais); 
+              if (state == 1)  digitalWrite(CONFIG_PIN1, state_relais);
```
Why not something like this:
```c++
state--;
pinMode(CONFIG_PIN1, OUTPUT);
if (state <= 1) digitalWrite(CONFIG_PIN1, state);
```

StefanNetzer replied:
I will check your proposal, and reply

StefanNetzer replied:
Done

StefanNetzer committed:
add events

Add additional events per schedule line

The clock - output plugin is an easy way to create an event scheduler. But, it does not provide events for each event, but only a generic on/off event.

With this change I have two values the schedule line # (variable schedule) and the event value (i.e. Temp).

I have changed the selector to an FormNumericBox, since this is more convienient to enter a value.

Sensor
GPIO → Clock Event: - None -
Day,Time 1: All,00:00
Temp1: 0
Day,Time 2: All,00:00
Temp2: 0

With this one gets two output Values

Output
Temp

and of cause two events, which I may evaluate in the rules

556164: EVENT: timer#Output=1.00
556261: EVENT: timer#Temp=10.00

With this enhancement it is easy to use ESP as an edge device for tons of use cases ( thermostat, irrigation systems )

StefanNetzer committed:
added TDs recommendations

added TDs recommendations

StefanNetzer committed:
Merge branch 'mega' of https://github.com/StefanNetzer/ESPEasy into mega

StefanNetzer committed:
cleanup

StefanNetzer committed:
resync to GitHub

StefanNetzer committed:
Update _P043_ClkOutput.ino

fixed typo

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -41,7 +42,7 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
       {
         strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[0], PSTR(PLUGIN_VALUENAME1_043));
-        strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], "Temp");
+        strcpy_P(ExtraTaskSettings.TaskDeviceValueNames[1], "Output2");
```
Please add it the same way as the VALUENAME1
By using PSTR macro (similar to F() macro) the string is not active in memory, but loaded from flash.
Using static strings in C++ code will have the  string content in memory twice, so that's why the flash string work-around was implemented.

StefanNetzer replied:
Hi, thank you for the detailed explanation, I really appreciate your support here. 

Implemented and tested

StefanNetzer committed:
added PLUGIN_VALUENAME2_043

added PLUGIN_VALUENAME2_043

StefanNetzer committed:
Merge branch 'mega' into mega

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -53,25 +55,20 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
+            String options[3];
+            options[0] = "";
+            options[1] = F("Off");
+            options[2] = F("On");
```
I do see 3 options here.
On/off I can understand but what does the "" mean?

Also, to what question/option is it the answer?

StefanNetzer replied:
This is the code from the original version. You have 8 lines to define a schedule. So if you only need 2 times they will be defined on and off for example. The remaining lines will have "" and will not be executed. I volunteer to write the documentation of this plugin as well ;-)

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -53,25 +55,20 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
+            byte choice = ExtraTaskSettings.TaskDevicePluginConfig[x];
+            addSelector(String(F("p043_state")) + (x), 3, options, NULL, NULL, choice);
+          }
+          else addFormNumericBox(String(F("Temp")) + (x + 1), String(F("p043_state")) + (x), ExtraTaskSettings.TaskDevicePluginConfig[x],-100,100 );
```
Again the "Temp".
It is using the same HTML id as the On/Off/"" selector, but it is still a bit unclear to me what its relation it has to the Clock Output.
You mentioned before it is indeed a temperature, so I would at least expect a unit of measure after the field (`addUnit`)
But I'm not yet convinced the label should have such a specific use case relation.

StefanNetzer replied:
Sorry, missed that. I will change it to "value", which is generic. Change is uploaded.

StefanNetzer committed:
Update _P043_ClkOutput.ino

Changed Temp to generic value.

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -68,7 +68,7 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
             addSelector(String(F("p043_state")) + (x), 3, options, NULL, NULL, choice);
           }
-          else addFormNumericBox(String(F("Temp")) + (x + 1), String(F("p043_state")) + (x), ExtraTaskSettings.TaskDevicePluginConfig[x],-100,100 );
+          else addFormNumericBox(String(F("Value")) + (x + 1), String(F("p043_state")) + (x), ExtraTaskSettings.TaskDevicePluginConfig[x],-100,100 );
```
Since it is no longer a temperature, why limit the value to less then the range of the value you store it in?

StefanNetzer replied:
You are absolutely right, restriction gone.

StefanNetzer committed:
Update _P043_ClkOutput.ino

eliminated value range limits

StefanNetzer committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega

TD-er commented:
Still to be done:
- Make an output selector for the different options
- Test bahavior with Domoticz (XOR with the previous todo)
- Test the changes.  (would be great if someone else could test it....)

StefanNetzer commented:
> Still to be done:
> 
> * Make an output selector for the different options
> * Test bahavior with Domoticz (XOR with the previous todo)
> * Test the changes.  (would be great if someone else could test it....)

StefanNetzer committed:
Update _P043_ClkOutput.ino

StefanNetzer commented:
I added the Output Data Type like in the Dummy device, to follow your recommendation with the output selector.

Device[deviceCount].OutputDataType = Output_Data_type_t::All;

Can I restrict the selection to single/dual, which would allow to switch between 1 and 2 Output values.

TD-er commented:
Why did you close this PR?

StefanNetzer commented:
Sorry, hit the wrong button :-(

StefanNetzer committed:
added output data type

added output data type

StefanNetzer committed:
replaced all by simple

StefanNetzer committed:
restored

StefanNetzer committed:
fixed link error

StefanNetzer committed:
using output value definition from  P004

StefanNetzer commented:
Output selector with type simple implemented.

StefanNetzer committed:
Merge branch 'mega' into mega

StefanNetzer committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega

StefanNetzer committed:
Merge branch 'mega' into mega

StefanNetzer commented:
Hi, did I miss anything. I think I have implemented all requests from you. Sorry for this newbie question ;-)

TD-er commented:
> Hi, did I miss anything. I think I have implemented all requests from you. Sorry for this newbie question ;-)

Well you 'forgot' to include one very important part in the PR... but that's a mistake lots of people make, so you're forgiven.
The thing you forgot to include is a bag filled with extra time to manage it all :)

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -53,25 +95,20 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
+            options[0] = "";
+            options[1] = F("Off");
+            options[2] = F("On");
+            addHtml(" ");
```
Please wrap this in a F-macro
```c++
addHtml(F(" "));
```

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -53,25 +95,20 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
-          addSelector(String(F("p043_state")) + (x), 3, options, NULL, NULL, choice);
+          if (CONFIG_PIN1 >= 0) {
+            String options[3];
+            options[0] = "";
```
You don't need to initialize a newly created `String` object with `""`.
Here it is clearly for readability, but I'm not entirely sure what the compiler does with this (unneeded) code.
The web interface isn't executed very often, so no need for commenting it out as optimization, so just as a remark to help in future PRs ;)

StefanNetzer replied:
done, added comment

TD-er commented on src/_P043_ClkOutput.ino:
```diff
@@ -112,12 +149,18 @@ boolean Plugin_043(byte function, struct EventStruct *event, String& string)
+              }
+              else {
+                UserVar[event->BaseVarIndex] = x+1;
+                UserVar[event->BaseVarIndex+1] = state;
```
Why is the order of task values changed if a GPIO is specified?
Also `Output_Data_type_t::Simple` makes it possible to set 3 or 4 output values, so what happens then?

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega

TD-er committed:
Merge branch 'mega' into mega
